tests/integration/utils/network.py

```diff
@@ -2,68 +2,71 @@
-        next_start = self.end + 1
-        next_end = next_start + number_of_ports - 1  # -1 because the range is inclusive
+        self.start = start
+        self.end = end
+        self._maybe_available = list(range(start, end))
 
+    def next_range(self, number_of_ports: int) -> "PortRangeInclusive":
+        next_start = self.end
+        next_end = next_start + number_of_ports
         return PortRangeInclusive(next_start, next_end)
 
-
-# To find a good port range use the following:
-#
-#   curl --silent 'https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt' | \
-#       egrep -i -e '^\s*[0-9]+-[0-9]+\s*unassigned' | \
-#       awk '{print $1}'
-#
-KAFKA_PORT_RANGE = PortRangeInclusive(48700, 48800)
-ZK_PORT_RANGE = KAFKA_PORT_RANGE.next_range(100)
-REGISTRY_PORT_RANGE = ZK_PORT_RANGE.next_range(100)
-TESTS_PORT_RANGE = REGISTRY_PORT_RANGE.next_range(100)
-
-
-def get_random_port(*, port_range: PortRangeInclusive, blacklist: List[int]) -> int:
-    """Find a random port in the range `PortRangeInclusive`.
-
-    Note:
-        This function is *not* aware of the ports currently open in the system,
-        the blacklist only prevents two services of the same type to randomly
-        get the same ports for *a single test run*.
-
-        Because of that, the port range should be chosen such that there is no
-        system service in the range. Also note that running two sessions of the
-        tests with the same range is not supported and will lead to flakiness.
-    """
-    assert port_range.start <= port_range.end, f"{port_range.start} must be less-than-or-equal to {port_range.end}"
-
-    # +1 because randint is inclusive for both ends
-    ports_in_range = (port_range.end - port_range.start) + 1
-    assert len(blacklist) < ports_in_range, f"no free ports available. Range {port_range}, blacklist: {blacklist}"
-
-    value = random.randint(port_range.start, port_range.end)
-    while value in blacklist:
-        value = random.randint(port_range.start, port_range.end)
-    return value
+    @contextmanager
+    def allocate_port(self) -> int:
+        """Find a random port in the range `PortRangeInclusive`.
+
+        Note:
+            This function is *not* aware of the ports currently open in the system,
+            the blacklist only prevents two services of the same type to randomly
+            get the same ports for *a single test run*.
+
+            Because of that, the port range should be chosen such that there is no
+            system service in the range. Also note that running two sessions of the
+            tests with the same range is not supported and will lead to flakiness.
+        """
+        if len(self._maybe_available) == 0:
+            raise RuntimeError(f"No free ports available. start: {self.start} end: {self.end}")
+
+        filtered_ports = ((pos, port) for pos, port in enumerate(self._maybe_available) if not is_time_wait(port))
```

Comment:
This was kinda annoying, at first I assumed that 300 sockets would be good enough, but then tests failed with `address already in use` errors. I thought it could be an issue with a port outside of the test run, so I added a check using `psutil` to make sure the port was not used, turn out the socket was on TIME_WAIT state. So I tried to change this to a FIFO queue, to give time for the socket to close, but that was not sufficient for every socket. So I changed it to search for the first socket that is not on TIME_WAIT state.

Now, the obvious suggestion here would be to do what @jjaakola-aiven mentioned above. I generally find _both_ of these solutions bad. But let me know what you think.